Dockerfile

```diff
@@ -1,37 +1,30 @@
-FROM golang:1.21 as build-stage
+# syntax=docker/dockerfile:1
 
-WORKDIR /fn
-
-COPY go.mod go.sum ./
-RUN go mod download
-
-COPY input/ ./input
-COPY *.go ./
-
-RUN CGO_ENABLED=0 go build -o /function .
-
-FROM debian:12.1-slim as package-stage
-
-# TODO(negz): Use a proper Crossplane package building tool. We're abusing the
-# fact that this image won't have an io.crossplane.pkg: base annotation. This
-# means Crossplane package manager will pull this entire ~100MB image, which
-# also happens to contain a valid Function runtime.
-# https://github.com/crossplane/crossplane/blob/v1.13.2/contributing/specifications/xpkg.md
-WORKDIR /package
-COPY package/ ./
+# We use the latest Go 1.x version unless asked to use something else.
+ARG GO_VERSION=1
 
-RUN cat crossplane.yaml > /package.yaml
-RUN cat input/*.yaml >> /package.yaml
+# Setup the base environment.
+FROM --platform=${BUILDPLATFORM} golang:${GO_VERSION} AS base
 
-FROM gcr.io/distroless/base-debian11 AS build-release-stage
+WORKDIR /fn
+ENV CGO_ENABLED=0
 
+COPY go.mod go.sum ./
+RUN --mount=type=cache,target=/go/pkg/mod go mod download
+
+# Build the Function.
+FROM base AS build
+ARG TARGETOS
+ARG TARGETARCH
+RUN --mount=target=. \
+    --mount=type=cache,target=/go/pkg/mod \
+    --mount=type=cache,target=/root/.cache/go-build \
+    GOOS=${TARGETOS} GOARCH=${TARGETARCH} go build -o /function .
+
+# Produce the Function image.
+FROM gcr.io/distroless/base-debian11 AS image
 WORKDIR /
-
-COPY --from=build-stage /function /function
-COPY --from=package-stage /package.yaml /package.yaml
-
+COPY --from=build /function /function
```

Comment:
the image does not need to contain `crossplane.yaml` anymore?

Reply:
The `crossplane xpkg build` command takes care of this now. It does it "better" in that it compiles `package.yaml` and puts it in a layer with a special annotation. That way the package manager can pull just the layer it needs to read `package.yaml`, not the whole (somewhat large) image. 

See https://github.com/crossplane/crossplane/blob/master/contributing/specifications/xpkg.md for details.